trin-storage/src/versioned/id_indexed_v1/pruning_strategy.rs

```diff
@@ -0,0 +1,360 @@
+use std::{fmt::Debug, ops::Range, time::Duration};
+
+use tracing::debug;
+
+use crate::versioned::usage_stats::UsageStats;
+
+use super::IdIndexedV1StoreConfig;
+
+/// The configuration parameters used by [PruningStrategy].
+#[derive(Clone, Debug)]
+pub struct PruningConfig {
+    /// The fraction of storage capacity that we aim for when pruning.
+    pub target_capacity_fraction: f64,
+    /// The fraction by which we increase/decrease the `max_pruning_count` when pruning duration is
+    /// outside `optimal_pruning_duration_range`.
+    pub max_pruning_count_change_fraction: f64,
+    /// The range of pruning durations that we consider optimal.
+    pub optimal_pruning_duration_range: Range<Duration>,
+}
+
+impl PruningConfig {
+    /// By default, we aim to prune down to 95% of storage capacity.
+    pub const DEFAULT_TARGET_CAPACITY_FRACTION: f64 = 0.95;
+    /// By default, we increase/decrease `max_pruning_count` by 20%.
+    pub const DEFAULT_CHANGE_FRACTION: f64 = 0.2;
+    /// By default, we consider optimal pruning duration between 0.1 and 0.3 seconds.
+    pub const DEFAULT_OPTIMAL_PRUNING_DURATION_RANGE: Range<Duration> =
+        Duration::from_millis(100)..Duration::from_millis(300);
+
+    pub fn new(
+        target_capacity_fraction: f64,
+        max_pruning_count_change_fraction: f64,
+        optimal_pruning_duration_range: Range<Duration>,
+    ) -> Self {
+        if !(0.0..1.0).contains(&target_capacity_fraction) {
```

Comment:
Nit: I suppose 1.0 is a valid choice here, which Range seems to exclude. Due to database compression, there was even a time when >1.0 choices might be valid. I haven't observed how SQL performs since the switchover. Anyway it all seems mostly moot since it's hard-coded for now, but just wanted to make a note that I think it would be fine to limit the input up to 2.0 or something, since it would just be about letting people play with configuration values eventually, I don't think it's a massive footgun that we need to panic to avoid letting them play with larger values.

Reply:
Definitely while coding I had a mental model that `target_capacity` should we lower than `storage_capacity`.

Maybe code would work even if that's not the case, but I would say it doesn't makes much sense, because both storage capacity and target capacity represent the raw content size (`content_id.len() + content_key.len() + content_value.len()`), or at least approximation (there is some extra storage used for indexes and other smaller columns). Irrelevant, but I don't think that there is any compression happening at the moment.

With that being said, I will allow value 1.0. And once we have real use case for using values higher than 1, we can come back to it (in which case we would probably have to rewrite some of the logic).